lib/Support/JSON.cpp

```diff
@@ -0,0 +1,140 @@
+//===- Json.cpp - Json Utilities --------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Utilities for file system path handling, supplementing the ones from
+// llvm::sys::path.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Support/JSON.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/ADT/StringSwitch.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace json = llvm::json;
+
+using namespace circt;
+using mlir::UnitAttr;
+
+static bool fitsInDouble(const APInt &apint) {
+  if (apint.getActiveBits() >= std::numeric_limits<int64_t>::digits)
+    return false;
+  int64_t value = apint.getSExtValue();
+  return value == static_cast<int64_t>(static_cast<double>(value));
+}
+
+LogicalResult circt::convertAttributeToJSON(llvm::json::OStream &json,
+                                            Attribute attr, Location loc) {
+  return TypeSwitch<Attribute, LogicalResult>(attr)
+      .Case<DictionaryAttr>([&](auto attr) {
+        json.objectBegin();
+        for (auto subAttr : attr) {
+          json.attributeBegin(subAttr.getName());
+          if (failed(convertAttributeToJSON(json, subAttr.getValue(), loc)))
+            return failure();
+          json.attributeEnd();
+        }
+        json.objectEnd();
+        return success();
+      })
+      .Case<ArrayAttr>([&](auto attr) {
+        json.arrayBegin();
+        for (auto subAttr : attr)
+          if (failed(convertAttributeToJSON(json, subAttr, loc)))
+            return failure();
+        json.arrayEnd();
+        return success();
+      })
+      .Case<BoolAttr>([&](auto attr) {
+        json.value(attr.getValue());
+        return success();
+      })
+      .Case<StringAttr>([&](auto attr) {
+        json.value(attr.getValue());
+        return success();
+      })
+      .Case<IntegerAttr>([&](auto attr) {
+        // If the integer can be accurately represented by a double, print
+        // it as an integer. Otherwise, convert it to an exact decimal string.
+        const auto &apint = attr.getValue();
+        if (fitsInDouble(apint)) {
+          json.value(apint.getSExtValue());
+        } else {
+          SmallVector<char, 16> buffer;
+          apint.toStringSigned(buffer);
+          json.value(buffer);
+        }
+        return success();
+      })
```

Comment:
I don't know if this is correct to do conversion to string as this will no longer round trip correctly. Also, I think JSON is ambiguous on the max size of an integer where that is implementation defined.

I'd prefer to keep integers as integers, doubles as doubles, and strings as strings.